packages/engine/src/simulation/engine.rs

```diff
@@ -91,42 +91,45 @@ impl Engine {
     /// and their outputs are merged into one Context object.
     async fn run_context_packages(&mut self) -> Result<()> {
         log::trace!("Starting run context packages stage");
+        // Need write access to state to prepare for context packages,
+        // so can't start state sync (with workers) yet.
         let (mut state, mut context) = self.store.take_upgraded()?;
         let snapshot = self.prepare_for_context_packages(&mut state, &mut context)?;
-        self.comms.state_snapshot_sync(&snapshot).await?; // Synchronize snapshot with workers
-        let pre_context = context.into_pre_context();
 
+        // Context packages use the snapshot and state packages use state.
+        // Context packages will be run before state packages, so start
+        // snapshot sync before state sync, so workers have more time to
+        // get the respective syncs done in parallel with packages.
+
+        // Synchronize snapshot with workers
+        self.comms.state_snapshot_sync(&snapshot).await?;
```

Comment:
I'm actually a little confused as to why we need to get confirmation of state_sync but we don't need confirmation of snapshot_sync. Any chance you could clarify a little in the docstring?

Do we not risk running context packages before state_snapshot_sync is finished? (In the same way state packages might have ran before state_sync finished)

Reply:
I'll add a comment.

The reason we need to wait for state sync is that state packages in the main loop shouldn't write to state before workers have finished reading state. In the case of the snapshot, the main loop also shouldn't write to the snapshot before the workers have finished reading it. But there's no risk of that happening, because the snapshot isn't written to at all after the start of the step, so we don't need confirmation of snapshot_sync.